mono/metadata/gc.c

```diff
@@ -700,6 +726,8 @@ ves_icall_System_GCHandle_GetTargetHandle (MonoObject *obj, guint32 handle, gint
 	case HANDLE_NORMAL:
 		return mono_gchandle_new (obj, FALSE);
 	case HANDLE_PINNED:
+		if (!is_object_pinnable (obj))
+			mono_set_pending_exception (mono_get_exception_argument ("value", "Object contains non-primitive or non-blittable data."));
```

Comment:
You need to actually return here (`-1` or whatever, doesn't matter - it gets discarded by the icall wrapper), or you'll allocate a handle one line below and then leak it. `mono_set_pending_exception` is not a no-return function like `mono_raise_exception`.